test/SchedulerTest.cpp

```diff
@@ -1,71 +1,80 @@
 #include "pipert/Scheduler.h"
 #include "gtest/gtest.h"
 
-#include <iostream>
-#include <utility>
-
 class Human {
 public:
   Human(const std::string& name, int age) : name_(name), age_(age) {}
 
   std::string GetName() const { return name_; }
   int GetAge() const { return age_; }
 
-private: 
+private:
   std::string name_;
   int age_;
 };
 
-class Printer {
-public:
-  void Print(pipert::PacketToProcess<Human> packet) {
-    std::cout << packet.data().GetName() << " - " <<  packet.data().GetAge() << std::endl;
-  }
-};
+using SchedulerDeathTest = pipert::Scheduler;
+
+template <class T>
+class SchedulerTest : public testing::Test {};
+
+using MyTypes = ::testing::Types<char, int, Human>;
+TYPED_TEST_SUITE(SchedulerTest, MyTypes, );
+
+void SchedulerInitTestWithParameter(int workers) {
+  pipert::Scheduler sch(workers);
+  exit (0);
+}
+
+void DoNothing() {
+  return;
+}
 
-/*
-TEST(SchedulerTest, SchedulerState) {
-    pipert::Scheduler sch1(0);
-    pipert::Scheduler sch2(2);
+TEST(SchedulerDeathTest, InitializationWithParameter) {
+#ifdef DEBUG
+  EXPECT_DEATH(SchedulerInitTestWithParameter(-5), ".*");
+  EXPECT_DEATH(SchedulerInitTestWithParameter(-1), ".*");
+#endif //DEBUG
 
-    EXPECT_EQ(sch1.IsRunning(), false);
-    EXPECT_EQ(sch2.IsRunning(), false);
+  EXPECT_EXIT(SchedulerInitTestWithParameter(0), ::testing::ExitedWithCode(0), ".*");
+  EXPECT_EXIT(SchedulerInitTestWithParameter(1), ::testing::ExitedWithCode(0), ".*");
+  EXPECT_EXIT(SchedulerInitTestWithParameter(20), ::testing::ExitedWithCode(0), ".*");
+}
 
-    sch1.Start();
-    EXPECT_EQ(sch1.IsRunning(), true);
-    EXPECT_EQ(sch2.IsRunning(), false);
+TEST(SchedulerDeathTest, InitializationWithoutParameter) {
+  EXPECT_EXIT({pipert::Scheduler sch; exit (0);}, ::testing::ExitedWithCode(0), ".*");
+}
 
-    sch1.Stop();
-    sch2.Start();
-    EXPECT_EQ(sch1.IsRunning(), false);
-    EXPECT_EQ(sch2.IsRunning(), true);
+TEST(SchedulerDeathTest, SchedulerDoubleStart) {
+  testing::FLAGS_gtest_death_test_style="threadsafe";
+  pipert::Scheduler sch;
 
-    sch1.Start();
-    sch2.Start();
-    EXPECT_EQ(sch1.IsRunning(), true);
-    EXPECT_EQ(sch2.IsRunning(), true);
+  sch.Start();
+  EXPECT_DEBUG_DEATH(sch.Start(), ".*");
+  sch.Stop();
+}
```

Comment:
The more I see this, the more I am sure we could put some avoidance into the Scheduler in case of doing something meaningful in release. See https://github.com/gerazo/pipert/issues/4
And than we could test for that. I don't think testing for assertions make sense as that behavior is already tested and protected by the assertion. So we do not need to test for it again. I would say, expect_death should be used only for real program crashes, not asserts.

Reply:
It's okay, if I can modify the implementation part here then I create tests for the new logic. What type of warning or functionality can be meaningful enough for this misuse of Start method?

Reply:
As I wrote above, it is useful to assert in debug, but it should still do something non-fatal.